File: WatchFramework/GameCreator.m
//
//  GameCreator.m
//  Watch Match
//
//  Created by Watch Set Team on 11/23/14.
//  Copyright (c) 2014 Watch Set Team. All rights reserved.
//

#import "GameCreator.h"

#import "GameTurn.h"

static const float kMatchPercent = .70;

@interface GameCreator ()

@end

@implementation GameCreator

static NSMutableArray *arrayOfSameNumbers(const NSUInteger count, const NSUInteger number) {
    NSMutableArray *result = [[NSMutableArray alloc] init];
    for (NSUInteger i = 0; i < count; i++) {
        [result addObject:@(number)];
    }
    return result;
}
    
static NSMutableArray *arrayOfRandomNumbers(const NSUInteger count, const NSUInteger numberUpperBound) {
    NSMutableArray *result = [[NSMutableArray alloc] init];
    
    BOOL isRandom = NO;
    
    while (!isRandom) {
        [result removeAllObjects];
        
        for (NSUInteger i = 0; i < count; i++) {
            [result addObject:@(arc4random_uniform((u_int32_t)numberUpperBound))];
            if (i > 0) {
                if (![result[0] isEqualToNumber:result.lastObject]) {
                    isRandom = YES;
                }
            }
        }
    }
    
    return result;
}

- (id)initWithNumberOfPlayers:(const NSUInteger)numberOfPlayers
                numberOfTurns:(const NSUInteger)numberOfTurns
               numberOfShapes:(const NSUInteger)numberOfShapes
             numberOfPatterns:(const NSUInteger)numberOfPatterns
                     criteria:(const GameCriteria)criteria {
    self = [super init];
    if (self != nil) {
        NSMutableArray *turns = [[NSMutableArray alloc] init];
        
        for (NSUInteger i = 0; i < numberOfTurns; i++) {
            const NSUInteger randomNumber = arc4random_uniform(100);
            
            NSMutableArray *patterns = [[NSMutableArray alloc] init];
            NSMutableArray *shapes = [[NSMutableArray alloc] init];
            
            if (randomNumber < kMatchPercent * 100) {
                const u_int32_t matchedShape = arc4random_uniform((u_int32_t)numberOfShapes);
                const u_int32_t matchedPattern = arc4random_uniform((u_int32_t)numberOfPatterns);
                
                switch (criteria) {
                    case MatchShapesCriteria: {
                        shapes = arrayOfSameNumbers(numberOfPlayers, matchedShape);
                        patterns = arrayOfRandomNumbers(numberOfPlayers, numberOfPatterns);
                    }
                        break;
                    case MatchPatternsCriteria: {
                        shapes = arrayOfRandomNumbers(numberOfPlayers, numberOfShapes);
                        patterns = arrayOfSameNumbers(numberOfPlayers, matchedPattern);
                    }
                        break;
                    case MatchShapesAndColorsCriteria: {
                        shapes = arrayOfSameNumbers(numberOfPlayers, matchedShape);
                        patterns = arrayOfSameNumbers(numberOfPlayers, matchedPattern);
                    }
                        break;
                }
            } else {
                shapes = arrayOfRandomNumbers(numberOfPlayers, numberOfShapes);
                patterns = arrayOfRandomNumbers(numberOfPlayers, numberOfPatterns);
            }
            
            GameTurn *turn = [[GameTurn alloc] initWithPatterns:patterns shapes:shapes criteria:criteria];
            
            [turns addObject:turn];
        }
        
        _turns = turns;
        
        _startTime = [NSDate dateWithTimeInterval:5 *60 sinceDate:[NSDate date]];
    }
    return self;
}

- (BOOL)writeToFile:(NSURL *)fileURL {
    NSMutableArray *turnDictionaries = [[NSMutableArray alloc] init];
    
    for (GameTurn *turn in self.turns) {
        [turnDictionaries addObject:turn.serializableDictionary];
    }
    
    NSDictionary *dictionary = @{@"turns" : turnDictionaries,
                                 @"time" : self.startTime};
    
    NSError *error;
    
    NSData *data = [NSPropertyListSerialization dataWithPropertyList:dictionary format:NSPropertyListXMLFormat_v1_0 options:0 error:&error];
    
    BOOL result = [data writeToURL:fileURL options:0 error:&error];
    if (!result) {
        NSLog(@"%@", error);
    }
    
    return result;
}

@end

File: WatchFramework/GameCreator.swift
//
//  GameCreator.swift
//  Watch Match
//
//  Created by Andrew Pontious on 11/26/14.
//  Copyright (c) 2014 TeamHand. All rights reserved.
//

import Foundation

public class GameCreator {
    let turns: [GameTurn]
    let startTime: NSDate
    
    let kMatchPercent = 0.70
    
    public init(numberOfPlayers: Int,
                numberOfTurns: Int,
                numberOfShapes: Int,
                numberOfPatterns: Int,
                criteria: GameCriteria) {
        
            var turns = [GameTurn]()
            
            for i in 0..<numberOfTurns {
                let randomNumber = Int(arc4random_uniform(100));
                
                var patterns = [Int]()
                var shapes = [Int]()
                
                if (randomNumber < Int(kMatchPercent * 100)) {
                    let matchedShape = Int(arc4random_uniform(UInt32(numberOfShapes)))
                    let matchedPattern = Int(arc4random_uniform(UInt32(numberOfPatterns)))
                    
                    switch (criteria) {
                        case GameCriteria.MatchShapesCriteria:
                            shapes = arrayOfSameNumbers(numberOfPlayers, matchedShape);
                            patterns = arrayOfRandomNumbers(numberOfPlayers, numberOfPatterns);
                            break;
                        case GameCriteria.MatchPatternsCriteria:
                            shapes = arrayOfRandomNumbers(numberOfPlayers, numberOfShapes);
                            patterns = arrayOfSameNumbers(numberOfPlayers, matchedPattern);
                            break;
                        case GameCriteria.MatchShapesAndColorsCriteria:
                            shapes = arrayOfSameNumbers(numberOfPlayers, matchedShape);
                            patterns = arrayOfSameNumbers(numberOfPlayers, matchedPattern);
                            break;
                    }
                } else {
                    shapes = arrayOfRandomNumbers(numberOfPlayers, numberOfShapes);
                    patterns = arrayOfRandomNumbers(numberOfPlayers, numberOfPatterns);
                }
                
                let turn = GameTurn(patterns:patterns, shapes:shapes, criteria:criteria)
                
                turns.append(turn)
            }
            
            self.turns = turns;
            
            self.startTime = NSDate(timeInterval:5 * 60, sinceDate:NSDate())
    }
    
    public func writeToFile(fileURL: NSURL) -> Bool {
        var turnDictionaries = NSMutableArray()
            
        for turn in self.turns {
            turnDictionaries.addObject(turn.serializableDictionary())
        }
        
        var dictionary = ["turns" : turnDictionaries,
                "time" : self.startTime];
            
        var error: NSError?
        
        let data = NSPropertyListSerialization.dataWithPropertyList(dictionary, format:NSPropertyListFormat.XMLFormat_v1_0, options:Int(0), error:&error)!
        
        let result = data.writeToURL(fileURL, options:NSDataWritingOptions(0), error:&error)
        if (!result) {
            println("(\result)")
        }
        
        return result;
    }
}

func arrayOfSameNumbers(count: Int, number: Int) -> [Int] {
    var result = [Int]()
    for i in 0..<count {
        result.append(number)
    }
    return result;
}

func arrayOfRandomNumbers(count: Int, numberUpperBound: Int) -> [Int] {
    var result = [Int]()
    
    var isRandom = false
    
    while (!isRandom) {
        result.removeAll()
        
        for i in 0..<count {
            result.append(Int(arc4random_uniform(UInt32(numberUpperBound))))
            if (i > 0) {
                if (result[0] != result.last) {
                    isRandom = true
                }
            }
        }
    }
    
    return result;
}
